combo-demo/player/Player.gd

```diff
@@ -0,0 +1,100 @@
+extends KinematicBody2D
+
+enum States { MOVE, LIGHT_ATTACK, HEAVY_ATTACK }
+
+const DRAG_FACTOR := 15.0
+const SPEED_MOVE := 400.0
+const SPEED_ATTACK := 1000.0
+
+var _state = States.MOVE
+var _accept_next_attack := true
+var _direction := Vector2.ZERO
+var _velocity := Vector2.ZERO
+
+var _previous_light_combo := "light_combo_3"
+var _previous_heavy_combo := "heavy_combo_2"
+
+onready var _skin := $Skin
+onready var _animation_player := $AnimationPlayer
+
+
+func _unhandled_input(event: InputEvent) -> void:
+	if event.is_action_pressed("light_attack") and _accept_next_attack:
+		_previous_heavy_combo = "heavy_combo_2"
+		
+		_state = States.LIGHT_ATTACK
+		_accept_next_attack = false
+		
+		_animation_player.stop()
+		_animation_player.play("idle")
+		
+		match _previous_light_combo:
+			"light_combo_1":
+				_animation_player.play("light_combo_2")
+			"light_combo_2":
+				_animation_player.play("light_combo_3")
+			"light_combo_3":
+				_animation_player.play("light_combo_1")
+		
+		_previous_light_combo = _animation_player.current_animation
+		return
+	
+	if event.is_action_pressed("heavy_attack") and _accept_next_attack:
+		_previous_light_combo = "light_combo_3"
+		
+		_state = States.HEAVY_ATTACK
+		_accept_next_attack = false
+		
+		_animation_player.stop()
+		_animation_player.play("idle")
+		
+		match _previous_heavy_combo:
+			"heavy_combo_1":
+				_animation_player.play("heavy_combo_2")
+			"heavy_combo_2":
+				_animation_player.play("heavy_combo_1")
+		
+		_previous_heavy_combo = _animation_player.current_animation
+		return
```

Comment:
This is the code I think we should change the most about this demo because it's very hard-coded.

It is completely fine in the context of a game where all you're going to have is one or two combos. But as we make these demos for educational purposes, we have to design them with what our students want to learn in mind.